### DISCUSSION TITLE
Can Java ONNX get GPU without try/catch?

### DISCUSSION THREAD

**Discussion by lidj22** [microsoft/onnxruntime#17836] [Other Q&A] [answered]
In the Java [docs](https://onnxruntime.ai/docs/get-started/with-java.html#run-on-a-gpu-or-with-another-provider-optional), we can add a CUDA GPU with the `addCUDA` method. 

In its implementation, it seems to first check via `extractCUDA` whether CUDA is available, then adds it if it is. However, the underlying methods live in `OnnxRuntime` which seem private.

Is there a way to expose `extractCUDA`, or is it not possible? Currently I think I'm handling this rather ungracefully, with 

```java
try {
    sessionOptions.addCUDA(0);
} catch (OrtException e) {
    sderr("an error occurred, using GPU instead");
}
```
I would prefer
```java
try {
    if (extractCUDA()) {
        sessionOptions.addCUDA(0);
    }
} catch (OrtException e) {
    sderr("an error occurred, using GPU instead");
}
```

May come in handy when I'm running a program with or without GPU; if it's intended to run on a CPU, it shouldn't raise errors about there being no GPU.

**Comment by Craigacp** [comment]
`extractCUDA` extracts the CUDA library, but it doesn't load it. Only adding the CUDA EP triggers library loading (which could fail if you don't have the right version of CUDA/cuDNN installed). So exposing it as a separate method wouldn't achieve your goal.

**Reply by lidj22** [reply]
Thanks, then is there a preferred builtin method to do it? Or would I need to write logic for that?

**Reply by Craigacp** [reply, marked as answer]
No, I think trying to add the CUDA EP and catching the exception is the way to check if it's available. Java has no insight into the system dynamic linker which is what makes the decision if it can run CUDA or not.

**Reply by lidj22** [reply]
Gotcha, thanks for the input!
